src/phongo_bson_encode.c

```diff
@@ -86,6 +90,49 @@ static int php_phongo_is_array_or_document(zval* val)
 	return IS_ARRAY;
 }
 
+/* Checks the return type of a bsonSerialize() method. Returns true on
+ * success; otherwise, throws an exception and returns false.
+ *
+ * TODO: obsolete once PHP 8.0+ is required and tentative return type is enforced.
+ */
+static inline bool phongo_check_bson_serialize_return_type(zval* retval, zend_class_entry* ce)
+{
+	if (instanceof_function(ce, php_phongo_persistable_ce)) {
+		// Instances of Persistable must return an array, stdClass, or MongoDB\BSON\Document
+		if (
+			Z_TYPE_P(retval) != IS_ARRAY && !(Z_TYPE_P(retval) == IS_OBJECT && (instanceof_function(Z_OBJCE_P(retval), zend_standard_class_def) || instanceof_function(Z_OBJCE_P(retval), php_phongo_document_ce)))) {
+			phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE,
+								   "Expected %s::%s() to return an array, stdClass, or %s, %s given",
+								   ZSTR_VAL(ce->name),
+								   BSON_SERIALIZE_FUNC_NAME,
+								   ZSTR_VAL(php_phongo_document_ce->name),
+								   PHONGO_ZVAL_CLASS_OR_TYPE_NAME(*retval));
+			return false;
+		}
+
+		return true;
+	}
+
+	if (instanceof_function(ce, php_phongo_serializable_ce)) {
+		// Instances of Serializable must return an array, stdClass, MongoDB\BSON\Document, or MongoDB\BSON\PackedArray
+		if (
+			Z_TYPE_P(retval) != IS_ARRAY && !(Z_TYPE_P(retval) == IS_OBJECT && (instanceof_function(Z_OBJCE_P(retval), zend_standard_class_def) || instanceof_function(Z_OBJCE_P(retval), php_phongo_document_ce) || instanceof_function(Z_OBJCE_P(retval), php_phongo_packedarray_ce)))) {
+			phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE,
+								   "Expected %s::%s() to return an array, stdClass, %s, or %s, %s given",
+								   ZSTR_VAL(ce->name),
+								   BSON_SERIALIZE_FUNC_NAME,
+								   ZSTR_VAL(php_phongo_document_ce->name),
+								   ZSTR_VAL(php_phongo_packedarray_ce->name),
+								   PHONGO_ZVAL_CLASS_OR_TYPE_NAME(*retval));
+			return false;
+		}
+
+		return true;
+	}
+
+	return false;
```

Comment:
Quoting https://github.com/mongodb/mongo-php-driver/pull/1449#discussion_r1280813934:

> Cconsider throwing PHONGO_ERROR_UNEXPECTED_VALUE since we shouldn't get here, but if you want to just return true here you can remove the true returns above and use if/else instead.

If we return false, we should absolutely throw to follow conventions. Otherwise, just return true and NOP.